damnit/backend/db_migrations.py

```diff
@@ -0,0 +1,181 @@
+            );
+            """
+        )
+    )
+
+
+def _to_v2(conn: sqlite3.Connection) -> None:
+    conn.execute("ALTER TABLE run_variables ADD COLUMN attributes")
+
+
+def _to_v3(conn: sqlite3.Connection) -> None:
+    conn.executescript(
+        dedent(
+            """
+            -- Tags related tables
+            CREATE TABLE IF NOT EXISTS tags(
+                id INTEGER PRIMARY KEY AUTOINCREMENT,
+                name TEXT UNIQUE NOT NULL
+            );
+            CREATE TABLE IF NOT EXISTS variable_tags(
+                variable_name TEXT NOT NULL,
+                tag_id INTEGER NOT NULL,
+                FOREIGN KEY (variable_name) REFERENCES variables(name) ON DELETE CASCADE,
+                FOREIGN KEY (tag_id) REFERENCES tags(id) ON DELETE CASCADE,
+                PRIMARY KEY (variable_name, tag_id)
+            );
+            """
+        )
+    )
+
+
+def _to_v4(conn: sqlite3.Connection) -> None:
+    conn.execute(
+        dedent(
+            """
+            -- Trigger to remove an orphaned tag after its last reference is deleted from variable_tags
+            CREATE TRIGGER IF NOT EXISTS delete_orphan_tags_after_variable_tag_delete
+            AFTER DELETE ON variable_tags
+            FOR EACH ROW
+            BEGIN
+                -- Check if the tag_id from the deleted row (OLD.tag_id)
+                -- no longer exists in any other row in variable_tags
+                DELETE FROM tags
+                WHERE id = OLD.tag_id
+                AND NOT EXISTS (
+                    SELECT 1 FROM variable_tags
+                    WHERE tag_id = OLD.tag_id
+                );
+            END;
+            """
+        )
+    )
+
+
+MIGRATIONS: list[Migration] = [
+    Migration(
+        to_version=1,
+        description="Baseline schema (core tables, views)",
+        apply=_to_v1,
+    ),
+    Migration(
+        to_version=2,
+        description="Add attributes column to run_variables",
+        apply=_to_v2,
+    ),
+    Migration(
+        to_version=3,
+        description="Add tags and variable_tags tables",
+        apply=_to_v3,
+    ),
+    Migration(
+        to_version=4,
+        description="Add trigger to cleanup orphaned tags",
+        apply=_to_v4,
+    ),
+]
+
+
+def pending_migrations(from_version: int, to_version: int) -> Iterable[Migration]:
+    for m in MIGRATIONS:
+        if from_version < m.to_version <= to_version:
+            yield m
+
+
+def apply_migrations(
+    conn: sqlite3.Connection,
+    from_version: int,
+    to_version: int,
+    set_version: Callable[[int], None],
+) -> list[Migration]:
+    """
+    Apply all migration steps to go from `from_version` up to `to_version`.
+    Sets the version after each successful step using `set_version`.
+    Returns the list of applied migrations.
+    """
+    applied: list[Migration] = []
+    for step in pending_migrations(from_version, to_version):
+        # Each step in its own transaction for clarity and resilience
```

Comment:
I'm not sure what's common for these things, but I'd be inclined to put them all in 1 transaction. If I try to migrate a v1 database to v4 and it errors out, I'd prefer to be left with the unchanged v1 database rather than a v2 or v3 database. 🤷